traits/trait_types.py

```diff
@@ -1712,115 +1837,121 @@ def __init__(
+              provided but *high* is provided, then *high* provides the
+              default. If neither *low* nor *high* is provided, the default
+              is ``0.0``.
+        exclude_low : bool, optional
+            If true, the lower bound of the range is excluded. Otherwise
+            it's included. The default is to include the lower bound.
+        exclude_high : bool, optional
+            If true, the upper bound of the range is excluded. Otherwise
+            it's included. The default is to include the upper bound.
+        clip_on_get : bool, optional
+            If true, the ``get`` operation for this trait clips the stored
+            value to lie in [low, high]. Otherwise, the stored value is
+            returned as-is, and may be out of range if either low or high
+            has changed since the value was validated.
+
+            In legacy mode, the default is to clip if either the low or
+            high is dynamic. In normal mode, the default is not to clip.
+            Use of this option is not recommended in new code.
+        value_trait : TraitType, optional
+            A trait type, or object convertible to a trait type, representing
+            the value type for the range. For a numeric range, Int() or Float()
+            would be typical. In new code, it's recommended that *value_trait*
+            always be provided.
+
+            If *value_trait* is not provided, the type is
+            inferred from *low*, *high* and *value*, provided at least one of
+            those has numeric type. If none of *low*, *high* and *value* has
+            numeric type, a value type of ``Float`` is assumed and a
+            ``DeprecationWarning`` is issued. This may become an error in a
+            future Traits release.
+        """
+        # The legacy mode is provided for backwards compatibility.
+        # For new code, it's recommended to supply value_trait
+        # or use IntRange or FloatRange. Related: enthought/traits#590.
+        legacy_mode = value_trait is None
+
+        if legacy_mode:
+            value, value_name, value_trait = _infer_range_value_trait(
+                low, high, value)
         else:
-            self.get, self.set, self.validate = self._get, self._set, None
-            self._vtype = None
-            self._type_desc = "a number"
+            value_name = None
 
-            if isinstance(high, six.string_types):
-                self._high_name = high = "object." + high
-            else:
-                self._vtype = type(high)
-            high = compile(str(high), "<string>", "eval")
+        value_trait = trait_from(value_trait)
 
-            if isinstance(low, six.string_types):
-                self._low_name = low = "object." + low
-            else:
-                self._vtype = type(low)
-            low = compile(str(low), "<string>", "eval")
+        low_name = low if isinstance(low, six.string_types) else None
```

Comment:
I'd recommend adding `low_name` and `high_name` parameters and only do this if we are in legacy mode.  Otherwise ranges based on string types would be difficult to define.

Reply:
Will do. Then we can add `StrRange` along with `IntRange` and `FloatRange`.

Reply:
Done